File: Binary Search/maximizeSweetness.js
const maximizeSweetness = (sweetness, k) => {
  const numberOfPeople = k + 1;

  const check = (minSweetness) => {
    let currentSweetness = 0;
    let slices = 0;
    for (let chunk of sweetness) {
      currentSweetness += chunk;
      if (currentSweetness >= minSweetness) {
        slices++;
        if (slices === numberOfPeople) return true;
        currentSweetness = 0;
      }
    }

    return false;
  }

  let left = Infinity, right = 0;
  for (let chunk of sweetness) {
    left = Math.min(chunk, left);
    right += chunk;
  }

  right = Math.floor(right / numberOfPeople);

  while (left < right) {
    const mid = Math.floor((left + right + 1) / 2);
    if (check(mid)) {
      left = mid;
    } else {
      right = mid - 1;
    }
  }

  return right;
}

const sweetness = [1, 2, 2, 1, 2, 2, 1, 2, 2],
  k = 2;

console.log(maximizeSweetness(sweetness, k));

File: Binary Search/MaximizeSweetness.java
public class MaximizeSweetness {
  int numOfPeople;

  public int maximizeSweetness(int[] sweetness, int k) {
    numOfPeople = k + 1;
    int left = Integer.MAX_VALUE;
    int right = 0;
    for (int chunk : sweetness) {
      left = Math.min(chunk, left);
      right += chunk;
    }
    right /= numOfPeople;

    while (left < right) {
      int mid = (left + right + 1) / 2;
      if (check(mid, sweetness)) {
        left = mid;
      } else {
        right = mid - 1;
      }
    }

    return right;
  }

  public boolean check(int minSweetness, int[]sweetness) {
    int currentSweetness = 0;
    int slices = 0;
    for (int chunk : sweetness) {
      currentSweetness += chunk;
      if (currentSweetness >= minSweetness) {
        slices++;
        currentSweetness = 0;
        if (slices == numOfPeople)
          return true;
      }
    }

    return false;
  }
}
